relay-server/src/actors/spooler/mod.rs

```diff
@@ -179,148 +176,105 @@ impl FromMessage<RemoveMany> for Buffer {
-                .connect_with(options)
-                .await
-                .map_err(BufferError::SetupFailed)?;
+    /// Returns the number of envelopes in the memory buffer.
+    fn count(&self) -> usize {
+        self.buffer.values().map(|v| v.len()).sum()
+    }
 
-            let spool_config = BufferSpoolConfig {
-                db,
-                max_disk_size: config.spool_envelopes_max_disk_size(),
-                max_memory_size: config.spool_envelopes_max_memory_size(),
-                is_disk_full: false,
-            };
+    /// Removes envelopes from the in-memory buffer.
+    fn remove(&mut self, keys: &BTreeSet<QueueKey>) -> usize {
+        let mut count = 0;
+        for key in keys {
+            let (current_count, current_size) = self.buffer.remove(key).map_or((0, 0), |k| {
+                (k.len(), k.into_iter().map(|k| k.estimated_size()).sum())
+            });
+            count += current_count;
+            self.used_memory -= current_size;
+        }
+        relay_statsd::metric!(
+            histogram(RelayHistograms::BufferEnvelopesMemory) = self.used_memory as f64
+        );
+
+        count
+    }
 
-            service.spool_config = Some(spool_config);
+    /// Dequeues the envelopes from the in-memory buffer and send them to provided `sender`.
+    fn dequeue(&mut self, keys: &Vec<QueueKey>, sender: mpsc::UnboundedSender<ManagedEnvelope>) {
+        for key in keys {
+            for value in self.buffer.remove(key).unwrap_or_default() {
+                self.used_memory -= value.estimated_size();
+                sender.send(value).ok();
+            }
         }
+        relay_statsd::metric!(
+            histogram(RelayHistograms::BufferEnvelopesMemory) = self.used_memory as f64
+        );
+    }
 
-        Ok(service)
+    /// Enqueues the envelope into in-memory buffer.
+    fn enqueue(&mut self, key: QueueKey, managed_envelope: ManagedEnvelope) {
+        self.used_memory += managed_envelope.estimated_size();
+        self.buffer.entry(key).or_default().push(managed_envelope);
+        relay_statsd::metric!(
+            histogram(RelayHistograms::BufferEnvelopesMemory) = self.used_memory as f64
+        );
     }
 
-    /// Estimates the db size by multiplying `page_count * page_size`.
-    async fn estimate_buffer_size(db: &Pool<Sqlite>) -> Result<i64, BufferError> {
-        let size: i64 = sql::current_size()
-            .fetch_one(db)
-            .await
-            .and_then(|r| r.try_get(0))
-            .map_err(BufferError::FileSizeReadFailed)?;
+    /// Returns `true` if the in-memory buffer is full, `false` otherwise.
+    fn is_full(&self) -> bool {
+        self.used_memory >= self.max_memory_size
+    }
+}
 
-        relay_statsd::metric!(histogram(RelayHistograms::BufferDiskSize) = size as u64);
-        Ok(size)
+/// The configuration which describes the on-disk [`BufferState`].
+#[derive(Debug)]
+struct OnDisk {
```

Comment:
Both buffers share some behavior and APIs. Have you considered creating a trait?

Reply:
What would the benefit of the trait be here? I do not think we need a defined interface right now. 
I also didn't want to enforce `async` on non-async functions, and write the trait with async functions, which also would make this code a bit more complex. 
But if there is a strong opinion that we have to have  an interface/trait, I can definitely do it.